Add toggle to control invalid character substitution in group names

bcoca opened the pull request:
- move group name sanitation to inventory itself instead of plugins
 - create new toggle to allow user to customize 'valid names' WARNING, this can break plays bad characters are allowed.
 
Allows for  ansible/ansible#40581 to be implemented in aws_ec2 plugin

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
inventory

bcoca committed:
against my better judgement

bcoca committed:
handle bad regex

AlanCoding commented:
Some initial testing shows that:

```
ANSIBLE_INVALID_GROUP_CHARS="^$" ansible-inventory -i aws_ec2.yaml --list --export
```

Does seem to restore the behavior of the script, but I have not yet checked through everything.

mattclay commented:
CI failure in multiple tests:

* sanity: https://app.shippable.com/github/ansible/ansible/runs/109309/2/tests
* unit: https://app.shippable.com/github/ansible/ansible/runs/109309/5/tests
* integration: https://app.shippable.com/github/ansible/ansible/runs/109309/42/console

AlanCoding commented:
I'm hitting a snag with this.

```
AWS_ACCESS_KEY_ID=[removed] AWS_SECRET_ACCESS_KEY=[removed] ansible-inventory -i contrib/inventory/ec2.py --list --export
 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible/contrib/inventory/ec2.py with script plugin: Could not find group i-0588e7a7989cd029f in inventory

 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible/contrib/inventory/ec2.py with ini plugin:
/Users/alancoding/Documents/repos/ansible/contrib/inventory/ec2.py:3: Error parsing host definition ''''': No closing quotation

 [WARNING]: Unable to parse /Users/alancoding/Documents/repos/ansible/contrib/inventory/ec2.py as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

{
    "_meta": {
        "hostvars": {}
    }, 
    "all": {
        "children": [
            "ungrouped"
        ]
    }
}
```

The above command works fine in the `devel` branch, and yes, it returns groups with `-` in them.

For the rule that groups shouldn't use these characters, there just needs to be a timeframe for users to make the necessary changes. The reason I was running into this is because I was attempting to switch to the inventory plugins, and found that this rule was enforced for inventory via the plugins, but was never enforced for inventory via the scripts.

For other changes that affect playbooks, you have depreciation timelines with error messages that publish a version when the old behavior will be phased out. Could we have something like that for this group naming rule?

s-hertel commented on lib/ansible/inventory/data.py:
```diff
@@ -158,6 +158,7 @@ def get_host(self, hostname):
     def add_group(self, group):
         ''' adds a group to inventory if not there already '''
 
+        group = to_safe_group_name(group)
```
Yeah, AlanCoding makes a good point. It seems like the conversion should stay where it was for the time being, with a deprecation warning in `Group.__init__`. You could fix the composed groups with hyphens bugs before the add_group there.

I like the finer grain of control this gives rather than a toggle.

bcoca replied:
i expected my change to show deficiencies in the existing plugins, we need to fix them instead of 'passing on' the responsibility to the plugin author or we would allow invalid names into the system that can blow up randomly later. It is better to break 'at origin'.

bcoca replied:
this was my mistake in the initial design as the 'core/common' code should always have been the point of enforcement.

s-hertel replied:
We don't have tests for inventory scripts, so it's not clear how many things this breaks.

bcoca replied:
i pushed 'fix' for script plugin, i can add tests (thought we had simple one)

s-hertel replied:
I meant that we don't have tests for specific scripts that may or may not be generating inventory with '-' in group names. There are a couple correctly sanitizing group names, a couple doing it inadequately, and a number that appear to never sanitize. ~Adding the deprecation period resolves my concern since now the scripts have forewarning before they break.~
Strike that, actually took a look at the last commit. I guess changing the default behavior is another way to fix that.

bcoca replied:
i'm mostly exploring what the fix should look like, i'm not sure the current way is best since it doesn't cover 3rd party inventory plugins (but i expect almost none exist currently)

bcoca commented:
@AlanCoding the problem is that the 'script' plugin is not enforcing the rule, it should have

bcoca commented:
@s-hertel agreed, that is why i created toggle in the first place, but was rejected and the vote was 'it should break on bad names'

bcoca committed:
ensure script plugin uses safe  names

bcoca commented:
@AlanCoding now script plugin will work with rule

bcoca commented on lib/ansible/plugins/inventory/script.py:
```diff
@@ -162,6 +162,7 @@ def parse(self, inventory, loader, path, cache=None):
 
     def _parse_group(self, group, data):
 
+        group = to_safe_group_names(group)
```
unsure if this is correct fix or should add_group/add_host return 'actual name used' as a cleaner api?

kdelee commented:
@bcoca checked out this branch in a local docker environment of awx to see how it would effect inventory updates -- that was at about 11am so might not have that last commit you just pushed.

At that point, could not run ec2 inventory updates at all in awx. I am pretty concerned about the impact for awx

```
RuntimeError: ansible-inventory failed (rc=1) with stdout:

stderr:
 [WARNING]:  * Failed to parse /awx_devel/awx/plugins/inventory/ec2.py with
script plugin: Could not find group us-east-1b in inventory

 [WARNING]: Unable to parse /awx_devel/awx/plugins/inventory/ec2.py as an
inventory source

ERROR! No inventory was parsed, please check your configuration and options.
```

AlanCoding commented on lib/ansible/plugins/inventory/script.py:
```diff
@@ -46,7 +46,7 @@
 from ansible.module_utils._text import to_native, to_text
 from ansible.module_utils.common._collections_compat import Mapping
-from ansible.plugins.inventory import BaseInventoryPlugin, Cacheable
+from ansible.plugins.inventory import BaseInventoryPlugin, Cacheable, to_safe_group_name
```
imported as singular `name`, but used as plural, `names`.

bcoca committed:
make add_group return proper name

bcoca commented:
@kdelee as i posted above, i expected that to blow up. My initial code was ensuring the correctness at 'central' level instead of relying on plugins. In my last commits I'm now starting to fix the plugins themselves

kdelee commented:
ok, will keep an eye out and see where this goes, did not percolate into my brain this was not done

AlanCoding commented:
With the minor change that I noted above, I was able to run the ec2 script through this.

This produces group names such as `us_west_1`, not `us-west-1` as it used to.

My intention was to obtain a way to run the ec2 inventory plugin such that it can obtain the group names that were produced by the script. Here, you are changing it so that the script will produce the sanitized group names. Even though we vendor script in our own source code in order to avoid breaking changes, this would still cause an unavoidable breaking change on the upgrade of Ansible, even if we continued to run the scripts and didn't update them.

What you're doing will affect everybody, so I'm not really the one to review. It also doesn't really solve the problem for us to offer a graceful upgrade path that doesn't break things without asking first.

s-hertel commented:
I'm still in favor of a deprecation period for scripts rather than breaking backwards compat by default.

bcoca commented:
@s-hertel that would involve still accepting bad group names in general?

wenottingham commented:
Why is a character whitelist at all useful, rather than just a boolean (if this needs to be an overall setting)

In any case, this whole thing is backwards. You're changing ansible to force people to rewrite/redo their inventory, in support of not breaking theoretical playbooks *they do not have anyway*.

bcoca commented:
@wenottingham the plugins already change the group names, this is an effort to marry both requriements a) that we don't accept invalid group names and b) that sometimes we do

s-hertel commented:
@bcoca Yes, accepting bad group names from scripts for a time if INVALID_GROUP_CHARS is not set, with a deprecation warning that INVALID_GROUP_CHARS should be used instead. I like that #52045 allows specific keys to be configured rather than globally. I wonder if these two solutions could be merged.

wenottingham commented:
> @wenottingham the plugins already change the group names, this is an effort to marry both requriements a) that we don't accept invalid group names and b) that sometimes we do

My point is a toggle that allows the user to say that the letter 'A' and number '3'  (and only them) are invalid characters in group names is complication that serves no apparent real-life purpose and I don't think anyone has asked for, as opposed to just "allow invalid characters or not".

s-hertel commented:
Are you suggesting no sanitization? I don't think it's all or nothing. The ec2 script sanitizing, albeit not to core's standards. Different scripts sanitize different. This allows compatibility with all of those standards.

bcoca commented:
actually, the default sanitation of the ec2.py inventory script is the same as we use now, you only need to change it if using a specific option to allow for - in group names.

s-hertel commented:
@bcoca default for that option is true

wenottingham commented:
I would think internal script sanitization can stay if it's always been there - we just add no more of that, and make "additional santiization in core inventory parsing" toggleable via a boolean. Letting clowns like me pass `INVALID_GROUP_CHARS=[AaNnSsIiBbLlEe]` is overkill.

AlanCoding commented:
It's not just the ec2 script, both azure and gce are significantly affected by this. They do not have a similar option in the scripts, and I don't think anyone is interested in developing it.

s-hertel commented:
We could have a hyphen-specific boolean option and if true the regex would be "[^A-Za-z0-9_-]" and if not "[^A-Za-z0-9_]", defaulting to false. There are scripts that don't sanitize their group names so a plugin using either of those options may still break, and that's the only option I see where a user providing the regex would fix it. But since there aren't tests for all of the scripts/samples of their output I have no idea whether there are some returning interesting group names without fixing problematic characters and if that would be necessary. One problem with the PR I liked above is that you can make group names that contain any illegal character, such as ':', while really only '-' should be permitted to have compatibility with ec2/gce/azure.

AlanCoding commented:
I can give more data samples if needed. The gce script, in particular, makes group names out of IP addresses, meaning that `.` are commonly used characters. I'm not aware of any `:` in group names, I'm not sure if they were removed by Ansible or scripts just never put them in.

bcoca commented:
we've had cases of `:` in group names, he have had to specifically filter it out in past

wenottingham commented:
>  The gce script, in particular, makes group names out of IP addresses, meaning that `.` are commonly used characters. I'm not aware of any `:` in group names, I'm not sure if they were removed by Ansible or scripts just never put them in.

If GCE makes groups out of IP addresses, `:` will happen as soon as they get enough IPv6 roundtuits.

AlanCoding commented:
I would like to throw out a position that would give a path forward for us. I still need to double-check this with a few people, so I will try to do that and update on Monday to give certainty from our side.

 - A new boolean setting (as suggested in prior comments), where group names will be automatically sanitized if true
 - Default this setting to false
 - If a group name contains the prohibited characters, but setting is false, issue warning
 - declare deprecation period after which setting will change to true
 - you could leave the constructed plugin logic as-is, but it's not important either way

If you did this, then we would

 - do a data migration to convert group names as soon as possible
 - run subsequent inventory imports with the (above mentioned) setting set to true
 - have our inventory import logic convert group names to deal with users using old Ansible versions which does not have your patch

There would actually be quite a big benefit to doing it generally, as opposed to plugin-by-plugin. That is a constructive idea, we just need a little bit of time to get user data into it without disruption.

AlanCoding commented:
Discussed with some others (@matburt @jbradberry @wenottingham @kdelee), and I have some details to update.

Discussion directly relevant to this PR:

 - A boolean setting would be better. We do not want to maintain the regex syntax that would be used.
 - Need for a deprecation period is up in the air. We need some time before this change goes live, but if the default behavior changes in 2.8.0, that may or may not be too fast for us. I do need to ask if you have intent to backport this for 2.7. It would be problematic either way, because we want to enable some plugins in 2.7, but don't want to change the script's behavior in 2.7.

For how we would deal with this change:

 - We will not do data migrations of user data
 - Instead, we would pass through control of this setting to the user, probably via a general toggle to obtain "deprecated" behavior in inventory updates
 - Users will opt-in via this toggle for old sources, new sources will default to it
 - For older versions, the config variable can be statically set to use the deprecated behavior

kdelee commented:
Tried this out again out and running into another different problem.
After changing:
```
plugins/inventory/script.py:165:        group = to_safe_group_names(group)
```
to `to_safe_group_name`

I tried to run a "plain-ol-update" in awx using the ec2 script and this branch of ansible with no filters and got:
```
RuntimeError: ansible-inventory failed (rc=1) with stdout:

stderr:
 [WARNING]:  * Failed to parse /awx_devel/awx/plugins/inventory/ec2.py with
script plugin: can't add group to itself

 [WARNING]: Unable to parse /awx_devel/awx/plugins/inventory/ec2.py as an
inventory source

ERROR! No inventory was parsed, please check your configuration and options.
```

I updated exception to tell me what group it was (this is my change, original does not tell you what group):
```
inventory/group.py:161:            raise Exception("can't add {} group to itself".format(group))
```

Then tried a few more times, ran into several different groups it was erroring on:

`script plugin: can't add zones group to itself`
`script plugin: can't add tags group to itself`
`script plugin: can't add images group to itself`
`script plugin: can't add instances group to itself`

kdelee commented:
What I'm understanding on the awx side is we really want a Boolean setting, not a regex.

bcoca commented:
@kdelee i was using a regex as ec2.py is not the only option that adds an invalid character

bcoca committed:
move to boolean with warning and deprecation

bcoca commented:
updated to use a boolean, now it warns on transform and gives deprecation warning when not

AlanCoding commented:
Tried with current branch and got

```
ERROR! Unexpected Exception, this is probably a bug: Required argument 'string' (pos 1) not found
the full traceback was:

Traceback (most recent call last):
  File "/Users/alancoding/Documents/repos/ansible/bin/ansible-inventory", line 110, in <module>
    exit_code = cli.run()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/cli/inventory.py", line 121, in run
    self.loader, self.inventory, self.vm = self._play_prereqs()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/cli/__init__.py", line 479, in _play_prereqs
    inventory = InventoryManager(loader=loader, sources=options['inventory'])
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 127, in __init__
    self._inventory = InventoryData()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/data.py", line 58, in __init__
    self.add_group(group)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/data.py", line 161, in add_group
    group = to_safe_group_name(group)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/group.py", line 45, in to_safe_group_name
    matched = _UNSAFE_GROUP.search()
TypeError: Required argument 'string' (pos 1) not found
```

This is passing `ec2.py` to `ansible-inventory` with a config file, access key, and all that other stuff via env vars.

bcoca committed:
forgot what to search

AlanCoding commented:
Refreshed and got

```
ERROR! Unexpected Exception, this is probably a bug: 'NoneType' object has no attribute 'groups'
the full traceback was:

Traceback (most recent call last):
  File "/Users/alancoding/Documents/repos/ansible/bin/ansible-inventory", line 110, in <module>
    exit_code = cli.run()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/cli/inventory.py", line 121, in run
    self.loader, self.inventory, self.vm = self._play_prereqs()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/cli/__init__.py", line 479, in _play_prereqs
    inventory = InventoryManager(loader=loader, sources=options['inventory'])
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 127, in __init__
    self._inventory = InventoryData()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/data.py", line 58, in __init__
    self.add_group(group)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/data.py", line 161, in add_group
    group = to_safe_group_name(group)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/group.py", line 46, in to_safe_group_name
    for match in set(matched.groups):
AttributeError: 'NoneType' object has no attribute 'groups'
```

bcoca committed:
also allow for no matches

AlanCoding commented:
fetched again, and I'm back to getting this

```
 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible/contrib/inventory/ec2.py with script plugin: global name 'to_safe_group_names' is not defined

  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/script.py", line 161, in parse
    raise AnsibleParserError(to_native(e))
```

bcoca committed:
ensure central transform/check

 also fix some exception text handling

AlanCoding commented:
fetched, this time, produced quite a novel type of error

```
 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible/contrib/inventory/ec2.py with script plugin: unknown encoding: i-0f932e4a2e4d50c43

  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/script.py", line 161, in parse
    raise AnsibleParserError(to_native(e))
```

Given the error, my config file referenced by `EC2_INI_PATH` environment variable may be relevant.

```ini
[ec2]
base_source_var = value_of_var
regions = us-east-2,ap-south-1
regions_exclude = us-gov-west-1,cn-north-1
destination_variable = public_dns_name
vpc_destination_variable = ip_address
route53 = False
all_instances = True
all_rds_instances = False
include_rds_clusters = False
rds = False
nested_groups = True
elasticache = False
stack_filters = False
instance_filters = foobaa
group_by_ami_id = False
group_by_availability_zone = True
group_by_aws_account = True
group_by_instance_id = False
group_by_instance_state = True
group_by_platform = True
group_by_instance_type = True
group_by_key_pair = True
group_by_region = True
group_by_security_group = True
group_by_tag_keys = False
group_by_tag_none = False
group_by_vpc_id = False
cache_path = /tmp/foo
cache_max_age = 300
replace_dash_in_groups = True
```

I removed the config file and got the same error, but when I do that it might cache the content. That may or may not be relevant.

AlanCoding commented:
I tried from scratch with a python3 virtualenv, and got this

```
 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible/contrib/inventory/ec2.py with script plugin: not enough arguments for format string

  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/script.py", line 161, in parse
    raise AnsibleParserError(to_native(e))
```

bcoca committed:
minor fixes

AlanCoding commented:
After fetching the last fix I still get those same errors in python2 / python3. I can get more info by modifying the source:

```diff
diff --git a/lib/ansible/plugins/inventory/script.py b/lib/ansible/plugins/inventory/script.py
index 861cc80bc7..8b0a1c126f 100644
--- a/lib/ansible/plugins/inventory/script.py
+++ b/lib/ansible/plugins/inventory/script.py
@@ -158,6 +158,7 @@ class InventoryModule(BaseInventoryPlugin, Cacheable):
                 self._populate_host_vars([host], got)
 
         except Exception as e:
+            raise
             raise AnsibleParserError(to_native(e))
 
     def _parse_group(self, group, data):
```

result:

```
 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible/contrib/inventory/ec2.py with script plugin: unknown encoding: i-0f932e4a2e4d50c43

  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/script.py", line 146, in parse
    self._parse_group(group, gdata)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/script.py", line 166, in _parse_group
    group = to_safe_group_name(group)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/__init__.py", line 44, in to_safe_group_name
    return original_safe(name)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/group.py", line 48, in to_safe_group_name
    display.deprecated('Ignoring invalid character(s) "%s" in group (%s), in the future this will be an error' % to_text(invalid_chars, to_text(name)),
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/module_utils/_text.py", line 257, in to_text
    return to_text(value, encoding, errors)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/module_utils/_text.py", line 235, in to_text
    return obj.decode(encoding, errors)
```

bcoca committed:
fixed missing )

bcoca committed:
fix indentation issues

bcoca committed:
preserve orig error

AlanCoding commented:
Obtained output

```
[DEPRECATION WARNING]: Ignoring invalid character(s) "[u'-', u'-']" in group (us-east-1f), in the future this will be an error. This feature will be removed in version 2.12. 
Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
```

It gives a message like this for every group in the inventory, as far as I can tell.

These appear to go to stderr, so that works. After that JSON comes in, so that looks good.

I tried with env var

```
ANSIBLE__TRANSFORM_INVALID_GROUP_CHARS=true
```

Got:

```
ansible-inventory 2.8.0.dev0
  config file = None
  configured module search path = [u'/Users/alancoding/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/alancoding/Documents/repos/ansible/lib/ansible
  executable location = /Users/alancoding/.virtualenvs/ansible/bin/ansible-inventory
  python version = 2.7.11 (default, Oct 17 2016, 14:59:40) [GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
No config file found; using defaults
ERROR! Unexpected Exception, this is probably a bug: an integer is required
the full traceback was:

Traceback (most recent call last):
  File "/Users/alancoding/Documents/repos/ansible/bin/ansible-inventory", line 110, in <module>
    exit_code = cli.run()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/cli/inventory.py", line 121, in run
    self.loader, self.inventory, self.vm = self._play_prereqs()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/cli/__init__.py", line 479, in _play_prereqs
    inventory = InventoryManager(loader=loader, sources=options['inventory'])
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 127, in __init__
    self._inventory = InventoryData()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/data.py", line 58, in __init__
    self.add_group(group)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/data.py", line 165, in add_group
    g = Group(group)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/group.py", line 62, in __init__
    self.name = to_safe_group_name(name)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/group.py", line 44, in to_safe_group_name
    name = _UNSAFE_GROUP.sub(replacer, name, replace_and_warn)
TypeError: an integer is required
```

AlanCoding commented:
This has advanced far enough that I can start making a statement about the original thing we were trying to solve. The use case that I'm interested in is doing:

```
ansible-inventory -i aws_ec2.yaml --list --export
```

When I run this command, without changing the setting value, I get the deprecation warnings, and I get group names like `us-east-2`.

That is the thing that we were originally trying to accomplish. I am fine with that default behavior. However, to prevent dragging this on longer, I need to check that it is okay with you and everyone else.

To really contextualize this, we have to remember that in version 2.7 that same command gave group names of `us_east_2`. For me, that's not a problem because I know I was never utilizing this in 2.7.

bcoca commented:
i might need to revert my changes to aws_ec2 inventory plugin

bcoca commented:
also .. does your test group have BOTH an normal - and an unicode - ?

bcoca committed:
restored behaviour

AlanCoding commented:
> also .. does your test group have BOTH an normal - and an unicode - ?

"My" group comes from ec2. Maybe you are asking if names of some stuff have non-ascii characters in them. I can test that...

```
(auth stuff) contrib/inventory/ec2.py > ec2_data.txt
```

then...

```python
>>> with open('ec2_data.txt', 'r') as f:
...   astr = f.read()
... 
>>> print(all(ord(c) < 128 for c in astr))
True
```

So no, there are no non-ascii names that the script discovered, so if the plugin consumed some non-ascii characters, it got it in some way unique way.

----

I fetched your branch again. Now this command runs:

```
ANSIBLE__TRANSFORM_INVALID_GROUP_CHARS=true ansible-inventory -i private/ec2/aws_ec2.yaml --list --export -vvvvv
```

I have a problem, though. That command gives group names like `us_west_1`. This command:

```
ansible-inventory -i private/ec2/aws_ec2.yaml --list --export -vvvvv
```

(same as)

```
ANSIBLE__TRANSFORM_INVALID_GROUP_CHARS=false ansible-inventory -i private/ec2/aws_ec2.yaml --list --export -vvvvv
```

_also_ gives group names like `us_west_1`.

If you look at https://github.com/ansible/ansible/pull/52045, I hope it's clear that we needed a toggle to allow unsafe / safe group names from inventory plugins. Do did not need a toggle for scripts, we just needed the behavior to not change so that we don't introduce new data migration problems.

The current behavior of this patch is that it does _not_ give us that toggle for plugins, which is what we need. I'm don't yet know if this is unintended, or if it's a miscommunication.

sivel commented on lib/ansible/config/base.yml:
```diff
@@ -1382,6 +1382,18 @@ HOST_PATTERN_MISMATCH:
+    - Make ansible transform invalid characters in group names supplied by inventory sources.
+    - If 'false' it will allow for the group name but warn about the issue.
+    - When 'true' it will replace any invalid charachters with '_' (underscore).
+  env: [{name: ANSIBLE__TRANSFORM_INVALID_GROUP_CHARS}]
```
This appears to have an extra errant `_` between `ANSIBLE` and `TRANSFORM`

bcoca commented:
@AlanCoding this is handling the 'central toggle', i restored the plugin behaviour to what it was before, otherwise it would be a breaking change.  We can look into additional toggles that are plugin specific.

bcoca committed:
fixes as per fb

bcoca committed:
added 'silent' to avoid spamming current users

  those already using the plugins were used to the transformations, so no need to alert them

kdelee commented:
@bcoca @jctanner @thaumos @wenottingham 

From our conversation the other day, I walked away with the understanding that we agreed on this:

> "We are going to add global config to disable group name sanitization so you can get your group names with dashes if you want them as well as whatever crazy thing you want. This will be implemented by doing group name sanitization at core level with global config to disable."

But it looks like what we've got here is:

> "Add setting to tell Ansible to sanitize group names at core. Another option is needed to sanitize group names at plugin level, but that's still TBD. Since the constructed plugin still sanitize the group names, we still don't get the dashes"

I understand the desire to have valid python variable names, and to have a centralized and standard sanitization method -- this is good. But since this sanitization is happening on multiple levels, we need the option to globally (at all levels) turn this off if we so deem necessary.

basically the problem for us is some of the plugins sanitize group names and we have no control of this behavior. We need dashes in our ec2 group names, users want it and use it even if it blocks them from doing certain actions with these group names. 

This setting does not effect that behavior so it does not solve our problem that caused us to file  https://github.com/ansible/ansible/issues/51844 

https://github.com/ansible/ansible/issues/40581 seems like the inverse of what we need

AlanCoding commented:
> #40581 seems like the inverse of what we need

Let us take that back. It was my fault, sorry I was giving some bad information. From that issue:

> like: ` - key: tags['Name'].replace('-', '_')` ?

is the opposite of what we want. We _just_ want a way to have the group names not be modified. All we want is `us-east-2` from the inventory plugin, even for a temporary amount of time so that users can change their stuff over at their own leisure.

kdelee commented:
oops yeah, what he said ^

bcoca commented:
@AlanCoding i understand what you want, but first im trying to establish a 'global' way to allow this, then the aws_ec2 plugin needs a diff toggle since the current behaviour doesn't do this transform

AlanCoding commented:
Sure, that is just saying that this doesn't address https://github.com/ansible/ansible/issues/40581 or https://github.com/ansible/ansible/issues/51844, and does something tangental to it. A change like https://github.com/ansible/ansible/pull/51958 is needed, by your wording:

> the aws_ec2 plugin needs a diff toggle

(it's not just aws_ec2, but the constructable plugin, but the solution could involved both)

that is needed without or without _this_ change. Could we get started on that as a separate PR?

bcoca commented:
@AlanCoding I'm just trying to get the 'general toggle' working with the system w/o breaking any current playbooks, after that we can start giving you the migration path.

I was not going to solve everything in one PR.

s-hertel commented on lib/ansible/config/base.yml:
```diff
@@ -1382,6 +1382,18 @@ HOST_PATTERN_MISMATCH:
+  default: False
+  description:
+    - Make ansible transform invalid characters in group names supplied by inventory sources.
+    - If 'false' it will allow for the group name but warn about the issue.
```
If a customer has any non "-" invalid characters in their inventory groups this won't provide backwards compatibility for the ec2/azure inventory plugins with the scripts, since the scripts would have replaced those non "-" invalid characters with underscores. Tags with invalid characters or IPv6 are possible examples.

The gce script doesn't sanitize, so a blanket toggle like this would keep the gce inventory plugin groups compatible

bcoca replied:
the default is false, but will give users a deprecation notice now, once we toggle to 'true' it will 'autotransform' with a warning for those plugins that don't already do the transform themselves

AlanCoding replied:
It sounds like you are assuming that some scripts like `ec2.py` sanitize group names. It has a setting for sanitizing group names, yes. But that doesn't mean that it works. I talked about this here:

https://github.com/ansible/ansible/issues/40581#issuecomment-467990027

I can file a bug if you'd like. I have found the behavior to be highly reproducible, and have dived in a little what values it has in code. It appears that `self.replace_dash_in_groups` does take the "True" value (meaning it should sanitize), but I think that either the regex was wrong, or the method `to_safe` was never called.

bcoca replied:
no, assuming the opposite, those using the script plugin will now get a deprecation warning when trying to pass invalid group names

AlanCoding replied:
@s-hertel I can safely say that in the `ec2.py` script, `to_safe` is not called on the relevant group names, and that the group name `us-east-2` is returned with either `replace_dash_in_groups = True` or `replace_dash_in_groups = False` (with the config loaded correctly, that is ruled out). I can get you more information on this if you want.

s-hertel replied:
@AlanCoding I'm not sure if you're understanding me. EC2 script *always* sanitizes, just to varying degrees. Here's my test case. Try adding a bad character in a tag. For instance, host has tags: {'bad:idea': 'here:too'}. Using the script on devel that creates the group `tag_bad_idea_here_too`. Those `:` are replaced by `_` (entirely regardless of replace_dash_in_groups). A filter is being used for bad characters, just not necessarily `-`. So now if you port to using the plugin, you're not going to be able to replicate that behavior while also creating groups with *some* invalid characters (i.e. `-`) because it's all or nothing, no partial sanitization (the group will become `tag_bad:idea:here:too` if you're using the toggle to create groups like `us-east-1`). Anyway, I don't use it and ripping off the bandaid is probably good. I'm not going to fight for it, just pointing out the deficiency and to have something to point to later if users run into the corner case while trying to replicate the behavior of the script via the plugin. Hopefully no one is doing that but... ¯\_ツ_/¯

s-hertel replied:
Related, a global solution like this also poses problems for combining different plugins per playbook run as you may want some plugins to filter and not others.

s-hertel replied:
@AlanCoding Ah, btw, replace_dash_in_groups is a [tag specific option](https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.ini#L122). Poorly named, doesn't work on general groups but works great for tags.

bcoca committed:
centralized valid var names

AlanCoding commented:
Given the updated information here, can I re-submit https://github.com/ansible/ansible/pull/52045 for consideration?

bcoca committed:
actually reverse, want invalid

bcoca commented:
@AlanCoding no, not removing the safety from constructed, but planning on adding a toggle to disable them

bcoca committed:
fixed regex

AlanCoding commented:
#52045 _is_ introducing a toggle, which makes me worried that something very important is not being communicated. From the examples of that PR:

```
keyed_groups
        - prefix: ""
          separator: ""
          key: ec2_region
          unsafe: true
```

The proposal is that if you don't use the "unsafe" key that you get the default behavior - which is to sanitize group names, which gives `us_east_1`, which is the way things are now. The above example would give `us-east-1`, returned from the plugin, before any scrubbing by the global inventory logic.

Your current PR here _already_ introduces a toggle. If I read correctly, you are saying that a _second_ toggle needs to be introduced to tell the plugins to sanitize group names.

If the toggle introduced in #52045 cannot be called a toggle for this purpose, then we must have completely different ideas about what problem needs to be solved.

bcoca commented:
@AlanCoding i'll look again, i must have misread the code as last time it looked like you completely removed the call

bcoca committed:
fix regex

bcoca committed:
just compile regex one time

bcoca committed:
dont display dupes

bcoca committed:
comment on regex

bcoca committed:
added regex tests

rockpenguin commented:
Hello!  Hopefully I am not too late to add a comment.  Tripped upon this thread due to using the Foreman plugin and having hyphens replaced in our group names.  I really like the idea of the "toggle", and my question is this: will the toggle stick around even after the deprecation?

We are using 'ansible-inventory' to export the hosts and groups from Foreman to use with Ansible.  I am running the output through a quick python script that formats into a yaml inventory suitable to our environment.  However, the Foreman plugin exports our Foreman/Puppet "hostgroups" as such: `tomcat/app-cep/development => tomcat_app_cep_development`.  This makes it hard to parse out that group name into its respective parts, versus something such as "tomcat_app-cep_development" where I can still parse it using the underscore as the delimiter into something like:

```yaml
appsvr1.example.com:
    platform: tomcat
    app_name: app-cep
    app_env: development
```

If the toggle remains going forward, then I can parse it into the parts I need, then replacing the hyphen with an underscore.  If the toggle were to go away, then I would have to continuously patch the Ansible code to produce group names with hyphens.  Unfortunately it would be an undertaking to convert our group names in Foreman/Puppet to use _ instead of -.  I can't imagine that we are the only ones affected by this given that hyphens are common in many of the systems to which Ansible can pull inventory.

Thanks!

AlanCoding commented:
@rockpenguin I have been doing my own testing of these kinds of differences, and I had not discovered what you mention because I have not yet gotten around to testing the foreman plugin. Ping @kdelee, he's been working on this too.

The foreman inventory plugin uses logic completely isolated from the common inventory plugin utilities.

https://github.com/ansible/ansible/blob/2033b3d066e092ca83e797e79afc3e2286656ea9/lib/ansible/plugins/inventory/foreman.py#L206

From reading this, I would expect the behavior that you describe - that `/` is replaced with `_`. This will be done for the foreman groups, which are unlike other plugins in that they are not user defined. I think these groups come from the hostvars `hostgroup_title` which comes from the server. Users can modify the group prefix, but cannot change which groups are produced.

The current state of this PR does two different things for different types of inventory plugins:

1. plugins that use the constructible base class and inputs like `keyed_groups` will retain their current behavior of sanitizing group names
2. _in the current form_ this PR removes the sanitization of group names from the foreman group plugin, which means that if you _want_ sanitization you will have to use the global toggle, and the default behavior will be to _not_ sanitize

The 2nd point sounds like something that might change soon in the future. You can see in the diff here that `original_safe(name, force=True, silent=True)` is called where the group sanitization used to be called (for the 1st point), meaning that it sanitizes group names regardless of the global setting. The same could be done (and would likely be done) for the foreman inventory plugin, in which case this would not help you get back the old group names, and some plugin-specific toggle would need to be added. In this case, https://github.com/ansible/ansible/pull/52045 would also not help you, because it is a plugin-specific toggle which applies to all plugins inheriting from constructible, of which foreman is not one of. So in this long-winded hypothetical, a foreman-specific patch would be needed to re-enable unsafe group names from the plugin (I would need this just like you).

I go through this long-winded hypothetical, because there seems to be a large amount of resistance to removing group name sanitization from an inventory type that has already applied it.

bcoca committed:
revert ini, script, foreman to prev behaviour

  ini and script will now warn about deprecation

bcoca committed:
more complete errormsg

s-hertel commented on lib/ansible/plugins/inventory/__init__.py:
```diff
@@ -318,6 +316,7 @@ def _add_host_to_composed_groups(self, groups, variables, host, strict=False):
         if groups and isinstance(groups, dict):
             self.templar.set_available_variables(variables)
             for group_name in groups:
+                group_name = to_safe_group_name(group_name)
```
If the group_name changes here the next line `groups[group_name]` is a KeyError.

bcoca replied:
he, missed that

s-hertel reviewed (commented):
Looks good to merge besides the minor composed groups thing.

bcoca committed:
move sanitization to avoid keyerror

AlanCoding commented:
I heard that the motivation for this was that group names should be valid variable names in python.

Someone made an interesting point to me, that in python3 unicode can be python variable names.

```
>>> 진서=4
>>> 진서
4
```

I thought I should raise this point, considering that all these will become a long series of underscores after this takes effect.

xarses commented:
I've created #56930 to help others find this, and the intent of the change; which appears to be missing from most of the documentation.
